Improve boxing of arguments to vararg methods

tpoliaw opened the pull request:
Potential fix for #100.

Only check final argument for a sequence if the correct number of arguments have been passed.

tpoliaw committed:
Improve boxing of arguments to vararg methods

If a Java method that accepts varargs is called from Jython, the
arguments are boxed into a PyList first. This should be skipped only if
there are already the correct number of arguments and the final
argument is a sequence type. If only the final argument is checked for
being a sequence it can block valid calls from matching the underlying
method.

eg in Java
    void foo(Object... args) {}
should be callable from Jython as
    foo(1, 2, [3, 4])
where the list and the numbers are all objects and passed as a single
array to the varargs method.

Similarly, if there are too few arguments, a java method such as
    void bar(int a, int[] b, int... c) {}
should be callable from Jython as
    bar(1, [2, 3])
where the vararg argument is empty and [2, 3] is passed as b.

The only remaining ambiguity is where a vararg parameter can take a
sequence type as an element, eg
    void fizz(Object... args) {}
calling from Jython as
    fizz([1,2,3])
would be valid as either args being [1,2,3] or [[1, 2, 3]]. This
ambiguity is also present in Java but behaves as if it was the former.

Stewori commented on tests/java/org/python/core/ReflectedArgsTest.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public class ReflectedArgsTest extends TestCase {
+	static class Demo {
+		public static void foo(Object... args) {}
```
Please use four spaces for indentation rather than tabs. (I am actually in favor of tabs in general, but one should stick to the project's style in this case)

tpoliaw replied:
Cheers. I matched the spaces in the other file but didn't think to check the new file.

tpoliaw commented:
Any chance of getting some feedback on this? It would be good to get this fixed without having to maintain the patched version locally.
Happy to rework the PR if required.

jeff5 commented:
@tpoliaw  Sorry, I've been having too much fun with Jython 3, but I'm having a push on Jython 2 issues now. I hope to accept one soon. It looks well-reasoned, and to have Java tests is reassuring.

Single-stepping through some of those will improve my understanding of argument matching. Then my opinion will be worth something!

I don't understand the "force push" on this PR. When I pull it to a branch on my repo I only see one change set (`9fb81f4`). However, if the difference is only white space I won't worry.

jeff5 committed:
Merge branch 'master' into pr-101

Experiment. Intend squash commit.

tpoliaw commented:
The force push was to update the commit with the whitespace changes  @Stewori requested rather than adding a new reformatting commit. The diff for before/after the update is [here](https://github.com/jython/jython/compare/c373c358f387714773b49c12e1d4cc08ccb8418f..9fb81f49a40e444fd300bc3e699fd072259c3fbc)

jeff5 commented:
That's what I guessed.

This looks good functionally, thanks. I'm merging in the tip of master for test, and will push that to your PR branch. Then there are some non-code changes (e.g. line length, and making your helpful commit message a comment) but I'll make and push those rather than ask for them.

I'll squash-merge at the end which should make a single change set in which all the extra change will have evaporated.

At least, I think that will work.

jeff5 committed:
Document and strengthen ReflectedArgsTest

Tweak code to match intent expressed in comments and examine the
argument array actually formed.

jeff5 commented:
@tpoliaw : I set out to give each test a Javadoc comment based on the comment in the body, but on close inspection, a couple of the the tests didn't actually do what they claim, so I adjusted the code to match the comment. That led me to want stronger tests of what actually comes back from `match` in  `callData` , and to have both `bar(int i, int[] j, int... rest)` and `baz(Integer i, Integer[] j, Integer... rest)`.

So the test is a lot longer now, and I understand what the code actually does. :)

Have I got this right and do they now cover the range of your addition to `ReflectedArgs`?

jeff5 commented:
And in a further confirmation, if `Demo` is made `public`, this all delivers the arguments one would hope:
```
Jython 2.7.3a1-DEV (heads/pr-101-merge:7b138dcaa, Mar 20 2022, 11:10:07)
[Java HotSpot(TM) 64-Bit Server VM (Oracle Corporation)] on java1.8.0_321
Type "help", "copyright", "credits" or "license" for more information.
>>> from org.python.core.ReflectedArgsTest import Demo
>>> Demo.foo(3, 4, ["bar"])
>>> Demo.foo([1,2,3])
>>> Demo.foo(1,2,3)
>>> Demo.foo("foo", "bar")
>>> Demo.bar(1, [2, 3])
>>> Demo.bar(1, [2, 3], 4, 5)
>>> Demo.bar(1, [2, 3], [4, 5])
>>> Demo.baz(1, [2, 3])
>>> Demo.baz(1, [2, 3], 4, 5)
>>> Demo.baz(1, [2, 3], [4, 5])
```

tpoliaw commented on tests/java/org/python/core/ReflectedArgsTest.java:
```diff
@@ -1,53 +1,167 @@
 
+    /**
+     * Calling {@code foo(Object... args)} as Python {@code foo("foo", "bar")} calls Java
+     * {@code foo("foo", "bar")}.
```
should this be `calls Java {@code foo(new String[] {"foo", "bar"})}`?

jeff5 replied:
I think I wrote `{@code foo(new Object[] {"foo", "bar"})}` first, but as that is quite hard to read and `foo` is a varargs in Java, I think what I have is ok.

tpoliaw commented on tests/java/org/python/core/ReflectedArgsTest.java:
```diff
@@ -1,53 +1,167 @@
 
+    /**
+     * Calling {@code foo(Object... args)} as Python {@code foo(3, 4, ["bar"])} calls Java
+     * {@code foo(3, 4, strlist})} where {@code strlist} is the representation of Python
```
looks like a mismatched `}` after `strlist`

jeff5 replied:
And here I wrote `{@code foo(new Object[] {3, 4, strlist})}` first (or something worse) then having reconsidered I didn't delete enough of it.

tpoliaw reviewed (commented):
Thanks for looking at this and improving the docs for the tests. I think everything looks ok to me. 
A couple of nitpicking comments on the javadoc but nothing that should block the PR.

jeff5 committed:
Minor changes to docs in test

jeff5 commented:
Sigh ... it fails irrelevantly (test_select), then succeeds on the rebound.
